web/core/components/workspace/sidebar/favorites/favorite-items/root.tsx

```diff
@@ -18,67 +24,139 @@ import {
     return combine(
       draggable({
         element,
         dragHandle: elementRef.current,
-        canDrag: () => true,
-        getInitialData: () => ({ id: favorite.id, type: "CHILD" }),
+        getInitialData: () => initialData,
         onDragStart: () => {
           setIsDragging(true);
         },
         onDrop: () => {
           setIsDragging(false);
         },
+        onGenerateDragPreview: ({ nativeSetDragImage }) => {
+          setCustomNativeDragPreview({
+            getOffset: pointerOutsideOfPreview({ x: "0px", y: "0px" }),
+            render: ({ container }) => {
+              const root = createRoot(container);
+              root.render(
+                <div className="rounded bg-custom-background-100 text-sm p-1 pr-2">
+                  <FavoriteItemTitle
+                    href={itemLink}
+                    icon={itemIcon}
+                    title={itemTitle}
+                    isSidebarCollapsed={!!sidebarCollapsed}
+                  />
+                </div>
+              );
+              return () => root.unmount();
+            },
+            nativeSetDragImage,
+          });
+        },
       }),
       dropTargetForElements({
         element,
+        canDrop: ({ source }) => getCanDrop(source, favorite, !!parentId),
         onDragStart: () => {
           setIsDragging(true);
         },
-        onDragEnter: () => {
-          setIsDragging(true);
+        getData: ({ input, element }) =>{
+
+          const blockedStates: InstructionType[] = ['make-child'];
+          if(!isLastChild){
+            blockedStates.push('reorder-below');
+          }
+
+          return attachInstruction(initialData,{
+            input,
+            element,
+            currentLevel: 1,
+            indentPerLevel: 0,
+            mode: isLastChild ? 'last-in-group' : 'standard',
+            block: blockedStates
+          })
+        },
+        onDrag: ({ self, source, location }) => {
+          const instruction = getInstructionFromPayload(self, source, location);
+          setInstruction(instruction);
         },
         onDragLeave: () => {
-          setIsDragging(false);
+          setInstruction(undefined);
         },
-        onDrop: ({ source }) => {
-          setIsDragging(false);
-          const sourceId = source?.data?.id as string | undefined;
-          if (!sourceId || !favoriteMap[sourceId].parent) return;
-          handleRemoveFromFavoritesFolder(sourceId);
+        onDrop: ({  source, location }) => {
+          setInstruction(undefined);
+          const dropTargets = location?.current?.dropTargets ?? []
+          if(!dropTargets || dropTargets.length <= 0) return;
+
+          const dropTarget = dropTargets.length > 1 ? dropTargets.find(target=>target?.data?.isChild) : dropTargets[0];
+
+          const dropTargetData = dropTarget?.data as TargetData;
+
+          if(!dropTarget || !dropTargetData) return;
```

Comment:
This logic looks complex and similar to the favorite folder, try to separate it out to a common folder

Reply:
Sperated it out to a common file